### PR TITLE
docs: Add CodeMeta JSON-LD Context File

### LINKED ISSUE #891: Add CodeMeta JSON-LD Context File
# Description

As part of an IRIS-HEP initiative to improve the citation quality/ease for all software it has been recommended by @danielskatz that projects:
- Track authors/contributors to software as it is developed
- Create a citation metadata file in the software repo and keep it up to date
- Link Github and Zenodo to get DOIs for major releases
- In the README and docs clearly says how to cite the software

`pyhf` already does all of this, as:
- [x] There is an [`AUTHORS` file](https://github.com/scikit-hep/pyhf/blob/31d34a0edd91593d75181cc3ad71563a52600793/AUTHORS) as well as a [listing of the authors in the docs](https://scikit-hep.org/pyhf/index.html#authors)
- [x] There is a [`.zenodo.json`](https://github.com/scikit-hep/pyhf/blob/2b0a1e2ba41a2aa1233511524e3aa78e696ead34/.zenodo.json) that formats Zenodo with the correct metadata and adds the ORCID of the authors
- [x] There is a Zenodo DOI that is kept in sync with our release schedule (almost) fully automatically (almost as it requires 1 button click). [DOI](https://doi.org/10.5281/zenodo.1169739)
- [x] There is a Citation section in both the [`README`](https://github.com/scikit-hep/pyhf/blob/2b0a1e2ba41a2aa1233511524e3aa78e696ead34/README.rst#citation) and the [docs webpage](https://scikit-hep.org/pyhf/citations.html) with the preferred citation

It has been pointed out though that we should also look into adding a [CodeMeta Project JSON-LD Context File](https://codemeta.github.io/jsonld/) as with it

> zenodo will read [it] and will be basically the same as [schema.org](http://schema.org/) so that this will help google index software, and tools be able to use the metadata file in [softwareheritage.org](http://softwareheritage.org/) to automatically build a citation for a particular commit.

(cc @gordonwatts as this might be of interest)

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #891 

Add `codemeta.json` generated by [`codemetapy`](https://github.com/proycon/codemetapy) (with some edits &mdash; if you just run `codemetap pyhf` then it results in thousands of lines as it creates the union of all the extras). This is also an [optional submission step](https://joss.readthedocs.io/en/latest/submitting.html#typical-paper-submission-flow) to JOSS

> (Optional) create a metadata file describing your software and include it in your repository. We provide a script that automates the generation of this metadata.

ReadTheDocs build: https://pyhf.readthedocs.io/en/docs-add-codemeta/development.html#context-files-and-archive-metadata

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add codemeta.json file generated from codemetapy
   - Control release number in codemeta.json with bumpversion
* Add codemetapy to 'develop' extra
* Add test to verify requirements in codemeta.json
* Update project description metadata across setup.cfg, .zenodo.json, and codemeta.json
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@danielskatz your feedback and review is welcome here too.

**Comment by danielskatz** [comment]
I'm not an expert on the contents of the file, but it looks reasonable

**Review by kratsg — COMMENTED** [review]
We need to provide a way to automatically update this file or we need to write down instructions on when this file gets updated.

**Comment by matthewfeickert** [comment]
> We need to provide a way to automatically update this file or we need to write down instructions on when this file gets updated.

Related to this is https://github.com/proycon/codemetapy/issues/4 and https://github.com/proycon/codemetapy/issues/5 (Also huge thanks to Maarten who has been quite helpful and responsive!)

File: codemeta.json [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            "familyName": "Feickert",
+            "email": "matthew.feickert@cern.ch"
+        },
+        {
+            "@id": "https://orcid.org/0000-0001-6616-3433",
+            "@type": "Person",
+            "givenName": "Giordon",
+            "familyName": "Stark",
+            "email": "gstark@cern.ch"
+        }
```
Comment (lukasheinrich) [inline codemeta.json:33]: add @cranmer

File: setup.cfg [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [metadata]
 name = pyhf
 version = 0.5.2
-description = (partial) pure Python HistFactory implementation
+description = pure-Python HistFactory implementation with tensors and autodiff
```
Comment (matthewfeickert) [inline setup.cfg:4]: @lukasheinrich @kratsg Maybe we should update our PyPI description at the same time? Seems that it isn't too much of a stretch to drop the partial at this point.

**Comment by matthewfeickert** [comment]
> We need to provide a way to automatically update this file or we need to write down instructions on when this file gets updated.

There's now a check in `.github/workflows/release_tests.yml` (as `codemetapy` relies on information published to PyPI) and instructions added to the developing docs on the website.

File: .github/workflows/release_tests.yml [hunk 3]
```diff
@@ -32,3 +32,8 @@ jobs:
     - name: Canary test public API
       run: |
         python -m pytest -r sx tests/test_public_api.py
+    - name: Verify requirements in codemeta.json
+      run: |
+        python -m pip install jq "codemetapy>=0.3.4"
+        codemetapy --no-extras pyhf > codemeta_generated.json
+        diff <(jq -S .softwareRequirements codemeta_generated.json) <(jq -S .softwareRequirements codemeta.json)
```
Comment (matthewfeickert) [inline .github/workflows/release_tests.yml:39]: Taken from https://stackoverflow.com/questions/31930041/using-jq-or-alternative-command-line-tools-to-compare-json-files

File: .zenodo.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-    "description": "pure-Python implementation of HistFactory models with tensors and autograd",
+    "description": "pure-Python implementation of HistFactory models with tensors and autodiff",
```
Comment (matthewfeickert) [inline .zenodo.json:2]: We should probably also decide on one description and just use that across `setup.cfg`, `.zenodo.json`, and `codemeta.json`.

**Review by kratsg — APPROVED** [review]

**Review by lukasheinrich — APPROVED** [review]

File: setup.cfg (hunk 2 above)
Comment (matthewfeickert) [reply setup.cfg:4]: Going with "pure-Python HistFactory implementation with tensors and autodiff"

**Pull request merged**